crates/zk-por-core/src/global.rs

```diff
@@ -85,32 +136,117 @@ impl GlobalMst {
     }
 
     /// `recursive_level` count from bottom to top; recursive_level = 1 means the bottom layer; increase whilve moving to the top.
-    pub fn set_recursive_hash(&mut self, recursive_level: u32, index: usize, hash: HashOut<F>) {
+    pub fn set_recursive_hash(&mut self, recursive_level: usize, index: usize, hash: HashOut<F>) {
         debug!(
             "set_recursive_hash, recursive_level: {:?}, index: {:?}, hash: {:?}",
             recursive_level, index, hash
         );
         let idx = self.get_recursive_global_index(recursive_level, index);
         self.inner[idx] = hash;
     }
+
+    pub fn is_integral(&self) -> bool {
```

Comment:
can log how long it takes for this check; we can do the check concurrently; we may leave for future optimization